### ISSUE TITLE
Step File does not show preview

### ISSUE THREAD

**Issue by SkipTam** [manyfold3d/manyfold#2571]
**Describe the bug**
Uploaded a Step file as model. There is no preview. When downloading file, it only opens the Browser with step instructions instead of downloading. Problem is the same on Mobile as 

**To Reproduce**
1. Upload Step File via Web interface
2. see no preview in model overview
3. click on model and download

**Expected behavior**
With STL file you see a preview and can be downloaded as a file.

**Screenshots**
![Screenshot_20240823_113702](https://github.com/user-attachments/assets/c0e4736b-c44f-4ff4-9d93-843cdf38dee6)
![Screenshot_20240823_113645](https://github.com/user-attachments/assets/0db427db-cc4a-480b-87d9-afb568e31d8e)
![Screenshot_20240823_113721](https://github.com/user-attachments/assets/1c49b18c-f448-4e01-b0d6-7b2875609154)

**Desktop (please complete the following information):**
 - OS: Linux Nixos stable 24.05
 - Browser Firefox and Brave 
 - Version FW: V129.0 Brave: 1.69.153

**Smartphone (please complete the following information):**
 - Device: Ipad Air 
 - OS: 17
 - Browser Firefox and Safari

**Additional context**
Not sure if this is a bug or feature request. Sorry if it categorized wrong

**Comment by Floppy** [comment]
Yeah, STEP doesn't have previews yet; it's always interesting to see what people actually use! I didn't really expect anyone to be using STEP 😃 

I'll change this to a feature request (although it should be clearer that preview isn't available rather than just being blank).

**Comment by SkipTam** [comment]
Thanks! 
Yeah have been trying out Step files since they are a lot more precise than stl files.
I now also have the problem that .stl files open in a new tab instead of downloading. Is this a config problem on my end?

**Comment by Floppy** [comment]
No, that's a known bug. Should get fixed soon!

**Comment by JakeAi** [comment]
> Yeah, STEP doesn't have previews yet; it's always interesting to see what people actually use! I didn't really expect anyone to be using STEP 😃
> 
> I'll change this to a feature request (although it should be clearer that preview isn't available rather than just being blank).

I only use STEP because curves are actual arcs on modern printers and slicers. No more multi-faceted arcs.

**Comment by radim-ek** [comment]
Same here. I stop using Manyfold, because I switch to STEP files.

**Comment by Floppy** [comment]
Problem is, I think, because STEP files are CAD-type files, with primitives, curves, etc, they're much more difficult to display in three.js which is mostly oriented around triangle meshes. It would need some conversion engine to triangulate the cad primitives for preview, which is the same sort of problem we have with things like FreeCAD files.

Most likely the solution to this is to have a background triangulation job run on the server side, and then deliver a triangle mesh or static image for preview render purposes.

**Comment by radim-ek** [comment]
And what about autoamtically convert to STL and display this? 
https://stackoverflow.com/questions/55714806/convert-step-file-type-to-stl

It can be done by a JOB:

If the HASH of the file does not exist, proceed.
Convert the file to STL and write the hash of the STEP file to the table.
If the HASH of the file is the same, do nothing.
If the HASH of the file differs, convert it to STL and write the new hash to the table.

**Comment by Floppy** [comment]
Yep, that's the approach we'll have to take. Thanks for the links, there might be something we can use!

**Comment by OlegEnot** [comment]
Greetings!
Having seen this issue, I would like to leave a link to the [WebCADViewer](https://github.com/cjsatuforc/WebCADViewer) project, apparently based largely on three.js and FreeCAD.
It may be possible to use a similar approach for web preview of CAD files (STEP and others)

**Comment by radim-ek** [comment]
Still not STEP files? I found a https://github.com/f3d-app/f3d
And .... because I need repository "viewer" I will wait or ... do myself :)

**Comment by Floppy** [comment]
As it happens, I've just added some code to the latest release which is heading in the direction of producing renderable derivatives for files we can't currently render; bear with me a bit longer, we might be nearly there :)

**Comment by lipusal** [comment]
Hello! Curious if there's been a chance to progress on this? I'm about to get started self-hosting Manyfold and STEP support would be fantastic =)

**Comment by Floppy** [comment]
Not yet, though we do now have funding sorted for better previews, so hopefully before too long we should have at least static image previews for STEP.

**Comment by emkayg** [comment]
This is great news. Seems like static previews and step previews might move out the "one day" category then? :)

**Comment by Floppy** [comment]
Yep, I'm about to totallyrewrite the roadmap too, as those categories are vastly out of date. 😁

**Comment by BladerunnerxRC** [comment]
Any progress on this?

**Comment by Floppy** [comment]
We do now use F3D for preview renders, but because of various library dependency issues, we don't have STEP in there yet. I'll go back, take another look and see if we can enable it yet.

**Comment by Floppy** [comment]
Good news, OpenCascade did get rebuilt with the new dependencies in Alpine (https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/99651), so we can now enable it in our F3D build, which means we should get STEP preview. I'll get it running; fingers crossed!

**Comment by Floppy** [comment]
Bingo!

<img width="352" height="488" alt="Image" src="https://github.com/user-attachments/assets/b5f59df1-fc81-4959-8e73-98548bd9ccaf" />

**Comment by BladerunnerxRC** [comment]
Am I doing something wrong? I redeployed my Manyfold stack to Docker using the v0.143.0 manyfold image. Rescanned STEP files are still not displaying step file previews. Is it actually not merged with v0.143.0 yet?

**Comment by Floppy** [comment]
@BladerunnerxRC are you using the linuxserver image? If so, that's actually still waiting on the feature being merged (https://github.com/linuxserver/docker-manyfold/pull/22) - it's a bit confusing, sorry. 

If you're using the official one, have you enabled '3D Renders" in the "File Derivatives" settings?

**Comment by kinghat** [comment]
> If you're using the official one, have you enabled '3D Renders" in the "File Derivatives" settings?

changing this worked for me. is the missing interactive version of steps a limitation of the F3D library? i feel like youve mentioned it before but cant remember.

**Comment by Floppy** [comment]
The current interactive renderer doesn't support STEP, but F3D does; we currently use that for the static previews, but I'm working on using it for the interactive view as well, so we'll get there :)
